server/src-lib/Data/Parser/JSONPath.hs

```diff
@@ -0,0 +1,51 @@
+module Data.Parser.JSONPath
+  ( parseJSONPath
+  , JSONPathElement(..)
+  ) where
+
+import           Control.Applicative  ((<|>))
+import           Data.Attoparsec.Text
+import           Prelude
+
+import qualified Data.Text            as T
+
+data JSONPathElement
+  = Key T.Text
+  | Index !Int
+  deriving (Show, Eq)
+
+firstPathElementChar :: Parser Char
+firstPathElementChar = letter
+  <|> satisfy (`elem` ("_$" :: String))
+  <?> "The first character must be a letter, an underscore (_) or a dollar sign ($)."
+
+pathElement :: Parser JSONPathElement
+pathElement = do
+  firstChar <- firstPathElementChar
+  remain <- many' ( letter
+                <|> digit
+                <|> satisfy (`elem` ("-_$" :: String))
+                )
+  return $ Key $ T.pack (firstChar:remain)
+
+childPathElement :: Parser JSONPathElement
+childPathElement = skip (== '.') *> pathElement
+
+indexPathElement :: Parser JSONPathElement
```

Comment:
`indexPathElement` should always parse an `Index` (`Int`). 
So I suggest,
```haskell
indexParser :: Parser Int
```
We should construct `JSONPathElement` in the function where this is being called. 

Similarly for `pathElement` function. 
```haskell
keyParser :: Parser T.Text
```
Note:-
A function signature should convey its behaviour. Both `indexPathElement` and `pathElement` gives `Parser JSONPathElement` which is not explicit.